### PR TITLE
Create a Module with alternative methods for Enumerable

### PR DESCRIPTION
**Pull request by carlosprietobarron**
Please review the Enumerable project branch "feature-branch"

### CODE REVIEW

**Comment by Anna-Myzukina** [comment]
# Status: [changes required]


### Good job with the milestone. Few steps and your project can be approved:
`Implementation details`

- [x] Set up linter properly. It should appear as shown below:


### General


 * Pull Request has meaningful title and a short description from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243643/)
screenshot 


* Comments: It’s better to delete commented code. If you need to review a previous version of your code, use git history from [Pathwright]()
screenshot 


#### Please check the ruby doc for these [methods](https://ruby-doc.org/core-2.6.5/Enumerable.html) behaviour. Methods to look out for are the following:

* my_each
* my_each_with_index
* my_select
* my_all?
* my_any?
* my_none?
* my_map
* my_inject

### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*
*Happy coding * <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

File: main.rb [hunk 1]
```diff
@@ -0,0 +1,179 @@
+module Enumerable
+  def my_each
+    i = 0
+    length.times do
+      yield (self[i])
+      i += 1
+    end
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: should have working `my_each` **(covered by automated tests)** identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* my_each returns an Enumerator if no block is given

              [1,2,3].each
              #=> #<Enumerator: [1, 2, 3]:each>

             [1,2,3].my_each
             #=> [1, 2, 3]

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+  def my_each_with_index
+    i = 0
+    length.times do
+      yield(self[i], i)
+      i += 1
+    end
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: should have working `my_each_with_index` **(covered by automated tests)** identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+      arr_ret = []
+      i = 0
+      length.times do
+        arr_ret.push(self[i]) if yield(self[i])
+        i += 1
+      end
+      arr_ret
+    end
+  end
+
```
Comment (Anna-Myzukina) [inline main.rb:35]: should have working `my_select` **(covered by automated tests)** identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+      # puts 'no block here'
+      retValue = true
+      length.times do
+        return false unless self[i]
+
+        i += 1
+      end
+      retValue
+    end
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: should have working `my_all?` **(covered by automated tests)** identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+      i = 0
+      length.times do
+        return true if self[i]
+
+        retValue = false
+        i += 1
+      end
+    end
+    retValue
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: should have working `my_any?` **(covered by automated tests)** identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+      i = 0
+      length.times do
+        return false if self[i]
+
+        retValue = true
+        i += 1
+      end
+    end
+    retValue
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: * [ ] should have working `my_none?` **(covered by automated tests)** identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
* [ ]

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+        arr_ret.push(yield(self[i]))
+        # puts arr_ret.inspect
+        i += 1
+      end
+      arr_ret
+    else
+      arr_ret
+    end
+  end
+
```
Comment (Anna-Myzukina) [inline main.rb:170]: should have working `my_map` **(covered by automated tests)** identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+      num = 0
+      i = 0
+      (length - 1).times do
+        num = yield(self[i], self[i + 1]) if i == 0
+        num = yield(num, self[i + 1]) if i > 0
+        i += 1
+      end
+    end
+    num
+  end
```
Comment (Anna-Myzukina) [inline main.rb]: should have working `my_inject` **(covered by automated tests)** identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

**Comment by carlosprietobarron** [comment]
I have re-programmed the modules and tested them with several test cases
the my_map method has 2 presentations
my_map that is the regular method
and 
my_map_proc that accepts optional proc as argument.

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+      if par.is_a? Class
+        my_each { |x| counter += 1 if x.is_a?(par) }
+        return counter
+      end
+      unless par.nil?
+        my_each { |x| counter += 1 if x == par }
+      end
+    end
+    counter
+  end
```
Comment (euqueme) [inline main.rb]: * [ ] should have working `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_count

- [ ] my_count counts the number of elements yielding a true value if a block is given
```ruby
[5, 4, 1].my_count { |x| x > 2 } == [5, 4, 1].count { |x| x > 2 } => 2
```
- Yours is giving the wrong answer
image

- Real count
image

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+    return to_enum unless block_given?
+
+    arr_ret = []
+    i = 0
+    length.times do
+      arr_ret.push(yield(self[i]))
+      i += 1
+    end
+    arr_ret
+  end
```
Comment (euqueme) [inline main.rb]: ### [OPTIONAL]

* [ ] should have working `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

#### my_map 
- [ ] in the specifications of the project in the [Odin project](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks) # 12, you were asked to modify your my_map method to take either a proc or a block. And if both a proc and a block are given, only execute the proc. The real map doesn't have this feature, but you were asked to have it so I strongly recommend working on it

Given 
```ruby
my_proc = Proc.new {|num| num > 10 }
```
Result
```ruby
puts [18, 22, 5, 6] .my_map(my_proc) {|num| num < 10 }  => true true false false
```

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+      total = first
+      i = 1
+      (a.length - 1).times do
+        total = yield(total, a[i]) if i < a.length
+        i += 1
+      end
+    end
+    total
+  end
+end
```
Comment (euqueme) [inline main.rb]: - [ ] should have working `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))

####  my_inject

- [ ] combines all elements of enum by applying a binary operation, specified by a block
```ruby
(1..5).my_inject(4) { |prod, n| prod * n } == (1..5).inject(4) { |prod, n| prod * n } => 480
```
- Yours is retrieving a TypeError
image

- Real inject
image

- [ ] When a symbol is provided as an argument, my_inject should combine the elements of the enum using that symbol as a named method.
```ruby
[2, 5, 6].my_inject(:+) == [2, 5, 6].inject(:+) => 13
```
- Yours is retrieving a TypeError
image

- Real inject
image

**Review by euqueme — CHANGES_REQUESTED** [review]
> ### Pull Request under the review by Maria Eugenia Queme P. (TSE)

### Hello @carlosprietobarron  :balloon: 

### Thank you for doing the changes requested by the previous reviewer :medal_sports: some of your Enumerable methods still need to work properly, so please address the comments I've left directly in your code :point_down: 

**Hint:** follow this [guide](https://ruby-doc.org/core-2.6.5/Enumerable.html), all the examples for each of the Enumerable methods should be able to work using your own Enumerable methods.

### That's it! :raised_hands: 

### Please :pray: do the changes :man_technologist: and submit back for a code review when you are done :memo: 

Happy Coding :rocket: 

Any questions? is better to ask me on slack **@ Maru**

PROJECT STATUS: REQUESTED CHANGES :vertical_traffic_light:

File: main.rb (hunk 10 above)
Comment (carlosprietobarron) [reply main.rb]: There were two "my_map" and my_map_proc the second one was the one that accepts proc as an argument

Comment (carlosprietobarron) [reply main.rb]: Now I changed the names the original my_map it's now my_map_old and "my_map_proc" it's "my_map"

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+      i = 0
+      length.times do
+        arr.push(yield(self[i]))
+        i += 1
+      end
+      arr
+    else
+      to_enum
+    end
+  end
```
Comment (shshamim63) [inline main.rb]: **_Optional_**
`my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
Based on the provided document our map should be fully dependent on the block. So I don't think that we need another method called my_map_proc. You can check the attached link.

File: main.rb (hunk 9 above)
Comment (shshamim63) [inline main.rb]: `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from Odin
 counts the number of elements yielding a true value if a block is given
```ruby
[1, 2, 3].my_count {|element| element > 1}    #should return 2
```

File: main.rb (hunk 11 above)
Comment (shshamim63) [inline main.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

File: main.rb [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+        i += 1
+      end
+    end
+    total
+  end
+end
+
+def multiply_els(arr = [])
+  block = proc { |i, n| i * n }
+  puts arr.my_inject(&block)
```
Comment (shshamim63) [inline main.rb]: We don't need to print it inside cause we are gonna call the method with a print. For example,
```ruby
arr = [1, 2, 3]
puts multiply_els(arr)
```

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :space_invader: Changes Required :space_invader: 
___
Hi @carlosprietobarron :wave:,
Awesome job. Almost near to finish the project. Just some more edge cases that you need to cover to get done with it. And believe me, those changes are really easy to cover. All you need to do just follow the instructions below and after updating your code check them with the provided sample. 
___
### :spades: Mandatory Changes :spades: 
___
1. Your my_count is not covering one edge case. Check the comment for more description.
2. my_inject method also needs some touchup. Check the comments to fix it.
___
### :clubs: Optional Changes :clubs: 
___
1. Based on the provided document you don't need my_map_proc method. So I will recommend you to remove it. Cause in one of your upcoming projects you will be asked to add tests. Then it will be redundant. 
2. You can update readme with the list of methods you have covered here. Cause it does not look descriptive enough. 
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Review by pelzolga123 — APPROVED** [review]
Approved.

**Comment by pelzolga123** [comment]
## Great  work, Congratulation :+1: Project Approved.
You did all changes required previously, all methods work now. Before you merge this project, please add some short description of this project to your README.md file. Then go ahead and merge to development.

_Goodluck in your next project_ *Happy coding*

**Pull request merged**